zilliqa/src/message.rs

```diff
@@ -32,18 +32,36 @@ pub struct Proposal {
     pub agg: Option<AggregateQc>,
     pub committee: Committee,
     pub transactions: Vec<SignedTransaction>,
+    pub opaque_transactions: Vec<Hash>,
 }
 
 impl Proposal {
-    pub fn from_parts(block: Block, transactions: Vec<SignedTransaction>) -> Self {
+    pub fn from_parts(block: Block, transactions: Vec<VerifiedTransaction>) -> Self {
+        Self::from_parts_and_hashes(
+            block,
+            transactions.into_iter().map(|tx| (tx.tx, tx.hash)).unzip(),
+        )
+    }
+
+    pub fn from_parts_and_hashes(
+        block: Block,
+        transactions: (Vec<SignedTransaction>, Vec<Hash>),
+    ) -> Self {
+        let provided_tx_hashes = HashSet::<Hash>::from_iter(transactions.1);
         Proposal {
             header: block.header,
             qc: block.qc,
             agg: block.agg,
             committee: block.committee,
-            transactions,
+            transactions: transactions.0,
+            opaque_transactions: block
+                .transactions
+                .into_iter()
+                .filter(|hash| !provided_tx_hashes.contains(hash))
```

Comment:
It filters out all the `Hash`es that are provided alongside `Transaction`s, thus leaving, as the field name suggests, only the opaque transactions that are listed (as hashes) in `block.transactions` but without the body being provided.

However good thing you asked this, because upon re-reading, I'm finding that my naming is quite terrible here. I'll try to clarify the code to be more self-explanatory.